crates/rspack_plugin_split_chunks_new/src/plugin/max_size.rs

```diff
@@ -0,0 +1,322 @@
+
+    let chunk_group_db = &compilation.chunk_group_by_ukey;
+
+    let chunks_to_be_split = compilation
+      .chunk_by_ukey
+      .values_mut()
+      .par_bridge()
+      .filter_map(|chunk| {
+        let max_size_setting = max_size_setting_map.get(&chunk.ukey);
+        tracing::trace!("max_size_setting: {max_size_setting:#?} for {:?}", chunk.ukey);
+
+        let min_size = max_size_setting
+          .map(|s| &s.min_size)
+          .unwrap_or(&fallback_cache_group.min_size);
+        let max_async_size = max_size_setting
+          .map(|s| &s.max_async_size)
+          .unwrap_or(&fallback_cache_group.max_async_size);
+        let max_initial_size: &SplitChunkSizes = max_size_setting
+          .map(|s| &s.max_initial_size)
+          .unwrap_or(&fallback_cache_group.max_initial_size);
+
+        if max_size_setting.is_none()
+          && !(fallback_cache_group.chunks_filter)(chunk, chunk_group_db)
+        {
+          tracing::debug!("Chunk({}) skips `maxSize` checking. Reason: max_size_setting.is_none() and chunks_filter is false", chunk.chunk_reasons.join("~"));
+          return None;
+        }
+
+        let mut allow_max_size = if chunk.is_only_initial(chunk_group_db) {
+          Cow::Borrowed(max_initial_size)
+        } else if chunk.can_be_initial(chunk_group_db) {
+          let mut sizes = SplitChunkSizes::empty();
+          sizes.combine_with(max_async_size, &f64::min);
+          sizes.combine_with(max_initial_size, &f64::min);
+          Cow::Owned(sizes)
+        } else {
+          Cow::Borrowed(max_async_size)
+        };
+
+        // Fast path
+        if allow_max_size.is_empty() {
+          tracing::debug!(
+            "Chunk({}) skips the `maxSize` checking. Reason: allow_max_size is empty",
+            chunk.chunk_reasons.join("~")
+          );
+          return None;
+        }
+
+        let mut is_invalid = false;
+        allow_max_size.iter().for_each(|(ty, ty_max_size)| {
+          if let Some(ty_min_size) = min_size.get(ty) {
+            if ty_min_size > ty_max_size {
+              is_invalid = true;
+              tracing::warn!(
+                "minSize({}) should not be bigger than maxSize({})",
+                ty_min_size,
+                ty_max_size
+              );
+            }
+          }
+        });
+        if is_invalid {
+          allow_max_size.to_mut().combine_with(min_size, &f64::max);
+        }
+
+        Some(ChunkToBeSplit {
+          allow_max_size,
+          min_size,
+          chunk: chunk.ukey,
+        })
+      })
+      .collect::<Vec<_>>();
+
+    let compilation = Mutex::new(compilation);
+
+    let infos = chunks_to_be_split
+      .into_par_iter()
+      .filter_map(|info| {
+        let ChunkToBeSplit {
+          chunk,
+          allow_max_size,
+          min_size,
+        } = &info;
+        let results = deterministic_grouping_for_modules(
+          &compilation.lock().expect("Should not panic"),
```

Comment:
Will this create additional overhead comparing to plain iterator?

Reply:
It's intensional. I want this code to be parallel-safe so I could refactor them to run in parallel later.